src/ring_theory/mv_polynomial/weighted_homogeneous.lean

```diff
@@ -0,0 +1,455 @@
+/-
+Copyright (c) 2022 María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández
+-/
+import algebra.graded_monoid
+import data.mv_polynomial.variables
+
+/-!
+# Weighted homogeneous polynomials
+
+It is possible to assign weights (in a commutative additive monoid `M`) to the variables of a
+multivariate polynomial ring, so that monomials of the ring then have a weighted degree with
+respect to the weights of the variables. The weights are represented by a function `w : σ → M`,
+where `σ` are the indeterminates.
+
+A multivariate polynomial `φ` is weighted homogeneous of weighted degree `m : M` if all monomials
+occuring in `φ` have the same weighted degree `m`.
+
+## Main definitions/lemmas
+
+* `weighted_total_degree' w φ` : the weighted total degree of a multivariate polynomial with respect
+to the weights `w`, taking values in `with_bot M`.
+
+* `weighted_total_degree w φ` : When `M` has a `⊥` element, we can define the weighted total degree
+of a multivariate polynomial as a function taking values in `M`.
+
+* `is_weighted_homogeneous w φ m`: a predicate that asserts that `φ` is weighted homogeneous
+of weighted degree `m` with respect to the weights `w`.
+
+* `weighted_homogeneous_submodule R w m`: the submodule of homogeneous polynomials
+of weighted degree `m`.
+
+* `weighted_homogeneous_component w m`: the additive morphism that projects polynomials
+onto their summand that is weighted homogeneous of degree `n` with respect to `w`.
+
+* `sum_weighted_homogeneous_component`: every polynomial is the sum of its weighted homogeneous
+components.
+-/
+
+noncomputable theory
+
+open_locale classical big_operators
+
+open set function finset finsupp add_monoid_algebra
+
+variables {R M : Type*} [comm_semiring R]
+
+namespace mv_polynomial
+variables {σ : Type*}
+
+section add_comm_monoid
+variables [add_comm_monoid M]
+
+-- TODO: move!
+/-- Rewrite a finsupp.sum as a finsum. -/
+lemma _root_.finsupp.sum_eq_finsum {ι M N : Type*} [has_zero M] [add_comm_monoid N]
+  (f : ι →₀ M) (g : ι → M → N) (hg : ∀ i, g i 0 = 0):
+  finsupp.sum f g = finsum (λ i, g i (f i)) :=
+begin
+  rw [finsum_eq_sum_of_support_subset, finsupp.sum],
+  intro i,
+  simp only [function.mem_support, ne.def, finset.mem_coe, finsupp.mem_support_iff],
+  intros h h',
+  rw ← h' at hg,
+  exact h (hg i),
+end
```

Comment:
This is not the correct file for this lemma. We don't need it for this PR anymore, should we try to move it to the correct file, or just delete it?

Reply:
My instinct would be: if it is supposed to be in a file which hasn't yet been ported, then move it; if it's supposed to be in a file which has been ported, then just delete it and don't draw attention to it ;-)